contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,786 @@
+        bytes calldata blsProof
+    ) external whenNotPaused validatorExist(msg.sender) {
+        require(_checkVoteAddress(newVoteAddress, blsProof), "INVALID_VOTE_ADDRESS");
+        require(_voteToOperator[newVoteAddress] == address(0), "DUPLICATE_VOTE_ADDRESS");
+
+        address operatorAddress = msg.sender;
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.updateTime + 1 days <= block.timestamp, "UPDATE_TOO_FREQUENTLY");
+
+        valInfo.voteAddress = newVoteAddress;
+        valInfo.updateTime = block.timestamp;
+        _voteToOperator[newVoteAddress] = operatorAddress;
+
+        emit VoteAddressEdited(operatorAddress, newVoteAddress);
+    }
+
+    function unjail(address operatorAddress) public whenNotPaused validatorExist(operatorAddress) {
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.jailed, "NOT_JAILED");
+
+        require(
+            IStakeCredit(valInfo.creditContract).getSelfDelegationBNB() >= minSelfDelegationBNB,
+            "NOT_ENOUGH_SELF_DELEGATION"
+        );
+        require(valInfo.jailUntil <= block.timestamp, "STILL_JAILED");
+
+        valInfo.jailed = false;
+        numOfJailed -= 1;
+        emit ValidatorUnjailed(operatorAddress);
+    }
+
+    function delegate(address operatorAddress) external payable whenNotPaused validatorExist(operatorAddress) {
+        uint256 bnbAmount = msg.value;
+        require(bnbAmount >= minDelegationBNBChange, "INVALID_DELEGATION_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+
+        if (valInfo.jailed) {
+            // only self delegation
+            require(delegator == operatorAddress, "ONLY_SELF_DELEGATION");
+        }
+
+        uint256 shares = IStakeCredit(valInfo.creditContract).delegate{ value: bnbAmount }(delegator);
+        emit Delegated(operatorAddress, delegator, shares, bnbAmount);
+
+        _sync(valInfo.creditContract, delegator);
+        IGovToken(GOV_TOKEN_ADDR).delegateVote(delegator, operatorAddress);
+    }
+
+    function undelegate(address operatorAddress, uint256 shares) public whenNotPaused validatorExist(operatorAddress) {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        uint256 bnbAmount = IStakeCredit(valInfo.creditContract).undelegate(delegator, shares);
+        emit Undelegated(operatorAddress, delegator, shares, bnbAmount);
+
+        if (delegator == operatorAddress) {
+            _checkValidatorSelfDelegation(operatorAddress);
+        }
+
+        _sync(valInfo.creditContract, delegator);
+    }
+
+    function redelegate(
+        address srcValidator,
+        address dstValidator,
+        uint256 shares
+    ) public whenNotPaused validatorExist(srcValidator) validatorExist(dstValidator) validatorNotJailed(dstValidator) {
```

Comment:
public ==> external

Reply:
I think `validatorNotJailed(dstValidator)` is not needed, because you have checked it inside the function.

Reply:
Done